stacker/plan2.py

```diff
@@ -0,0 +1,345 @@
+
+    def run(self):
+        """Runs this step until it has completed successfully, or been
+        skipped.
+        """
+
+        watcher = None
+        if self.watch_func:
+            watcher = multiprocessing.Process(
+                target=self.watch_func,
+                args=(self.stack,)
+            )
+            watcher.start()
+
+        try:
+            if self.status_changed_func:
+                self.status_changed_func()
+
+            while not self.done:
+                self._run_once()
+        finally:
+            if watcher and watcher.is_alive():
+                watcher.terminate()
+                watcher.join()
+        return self.ok
+
+    def _run_once(self):
+        status = self.fn(self.stack, status=self.status)
+        self.set_status(status)
+        return status
+
+    @property
+    def name(self):
+        return self.stack.fqn
+
+    @property
+    def short_name(self):
+        return self.stack.name
+
+    @property
+    def requires(self):
+        return self.stack.requires
+
+    @property
+    def completed(self):
+        """Returns True if the step is in a COMPLETE state."""
+        return self.status == COMPLETE
+
+    @property
+    def skipped(self):
+        """Returns True if the step is in a SKIPPED state."""
+        return self.status == SKIPPED
+
+    @property
+    def failed(self):
+        """Returns True if the step is in a FAILED state."""
+        return self.status == FAILED
+
+    @property
+    def done(self):
+        """Returns True if the step is finished (either COMPLETE, SKIPPED or FAILED)
+        """
+        return self.completed or self.skipped or self.failed
+
+    @property
+    def ok(self):
+        """Returns True if the step is finished (either COMPLETE or SKIPPED)"""
+        return self.completed or self.skipped
+
+    @property
+    def submitted(self):
+        """Returns True if the step is SUBMITTED, COMPLETE, or SKIPPED."""
+        return self.status >= SUBMITTED
+
+    def set_status(self, status):
+        """Sets the current step's status.
+        Args:
+            status (:class:`Status <Status>` object): The status to set the
+                step to.
+        """
+        if status is not self.status:
+            logger.debug("Setting %s state to %s.", self.stack.name,
+                         status.name)
+            self.status = status
+            self.last_updated = time.time()
+            if self.status_changed_func:
+                self.status_changed_func()
+
+    def complete(self):
+        """A shortcut for set_status(COMPLETE)"""
+        self.set_status(COMPLETE)
+
+    def skip(self):
+        """A shortcut for set_status(SKIPPED)"""
+        self.set_status(SKIPPED)
+
+    def submit(self):
+        """A shortcut for set_status(SUBMITTED)"""
+        self.set_status(SUBMITTED)
+
+
+def build_plan(description=None, steps=None,
+               step_names=None, reverse=False):
+    graph = build_graph(steps)
```

Comment:
I can't think of a scenario where that could actually happen from the `stacker` CLI, since there's config validation to ensure that stacks isn't empty, but I'll add a test for that here.